matrix-client/src/Network/Matrix/Client.hs

```diff
@@ -203,16 +227,252 @@ getTokenOwner session =
+    
+newtype StateKey = StateKey T.Text
+  deriving stock Show
+  deriving newtype FromJSON
+
+newtype EventType = EventType T.Text
+  deriving stock Show
+  deriving newtype FromJSON
+
+data MRCreate = MRCreate { mrcCreator :: UserID, mrcRoomVersion :: Integer }
+  deriving Show
+
+instance FromJSON MRCreate where
+  parseJSON = withObject "RoomCreate" $ \o -> do
+    mrcCreator <- o .: "creator"
+    mrcRoomVersion <- o .: "room_version"
+    pure $ MRCreate {..}
+
+newtype MRName = MRName { mrnName :: T.Text }
+  deriving Show
+
+instance FromJSON MRName where
+  parseJSON = withObject "RoomName" $ \o ->
+    MRName <$> (o .: "name")
+
+newtype MRCanonicalAlias = MRCanonicalAlias { mrcAlias :: T.Text }
+  deriving Show
 
+instance FromJSON MRCanonicalAlias where
+  parseJSON = withObject "RoomCanonicalAlias" $ \o ->
+    MRCanonicalAlias <$> (o .: "alias")
+
+newtype MRGuestAccess = MRGuestAccess { mrGuestAccess :: T.Text }
+  deriving Show
+
+instance FromJSON MRGuestAccess where
+  parseJSON = withObject "GuestAccess" $ \o ->
+    MRGuestAccess <$> (o .: "guest_access")
+
+newtype MRHistoryVisibility = MRHistoryVisibility { mrHistoryVisibility :: T.Text }
+  deriving Show
+
+instance FromJSON MRHistoryVisibility where
+  parseJSON = withObject "HistoryVisibility" $ \o ->
+    MRHistoryVisibility <$> (o .: "history_visibility")
+
+newtype MRTopic = MRTopic { mrTopic :: T.Text }
+  deriving Show
+
+instance FromJSON MRTopic where
+  parseJSON = withObject "RoomTopic" $ \o ->
+    MRTopic <$> (o .: "topic")
+    
+data StateContent =
+    StRoomCreate MRCreate
+ -- | StRoomMember MRMember
+ -- | StRoomPowerLevels MRPowerLevels
+ -- | StRoomJoinRules MRJoinRules
+  | StRoomCanonicalAlias MRCanonicalAlias
+  | StRoomGuestAccess MRGuestAccess
+  | StRoomHistoryVisibility MRHistoryVisibility
+  | StRoomName MRName
+  | StRoomTopic MRTopic
+  | StOther Value
+ --- | StSpaceParent MRSpaceParent
+  deriving Show
+
+pStRoomCreate :: Value -> Parser StateContent
+pStRoomCreate v = StRoomCreate <$> parseJSON v
+
+pStRoomCanonicAlias :: Value -> Parser StateContent
+pStRoomCanonicAlias v = StRoomCanonicalAlias <$> parseJSON v
+
+pStRoomGuestAccess :: Value -> Parser StateContent
+pStRoomGuestAccess v = StRoomGuestAccess <$> parseJSON v
+
+pStRoomHistoryVisibility :: Value -> Parser StateContent
+pStRoomHistoryVisibility v = StRoomHistoryVisibility <$> parseJSON v
+
+pStRoomName :: Value -> Parser StateContent
+pStRoomName v = StRoomName <$> parseJSON v
+
+pStRoomTopic :: Value -> Parser StateContent
+pStRoomTopic v = StRoomTopic <$> parseJSON v
+
+pStRoomOther :: Value -> Parser StateContent
+pStRoomOther v = StOther <$> parseJSON v
+    
+instance FromJSON StateContent where
+  parseJSON v = 
+        pStRoomCreate v 
+    <|> pStRoomCanonicAlias v
+    <|> pStRoomGuestAccess v
+    <|> pStRoomHistoryVisibility v
+    <|> pStRoomName v
+    <|> pStRoomTopic v
+    <|> pStRoomOther v
+
+-- TODO(SOLOMON): Should This constructor be in 'Event'?
```

Comment:
That's a good question, perhaps we should keep the current `Matrix.Events` module for message events. I t looks like `StateEvent` and `StateContent` are metadata related. I now see what you mean by module organization :-) I can't tell without having a deeper look, but I would bundle the data types and their instances based on their usage or the origin api path, e.g. `Matrix.Message.Event` and `Matrix.Room.Event`, or something like that.

Though it would be great if everything can be exported through `Network.Matrix`, so that the user does not have to import too many modules. Thus we might want to change the existing `Event` data type to `MessageEvent`, or something like that.

Thus I don't think that matter for this PR.

Reply:
Rooms in Matrix are essentially json key-value stores. You can actually insert your own arbitrary data. The `StateContent` stuff allows you to directly interact with room state. This has some fancy uses but also lets you do things like have a bot modify the room topic and fetch the canonical room alias.

I'm still trying to wrap my head around the overall structure of the Matrix API, but considering this library is a 'low level' wrapper on it, I think we should try to use the Matrix schemas as a guide for the module and type hierarchies.